Upgrade to Play 2.5

mkurz opened the pull request:
@mumoshu Could you please review?

cw-kuoka commented on project/Build.scala:
```diff
@@ -24,7 +24,7 @@ object ApplicationBuild extends Build {
 
   def isPlay24: Boolean = {
-    playMajorAndMinorVersion == "2.4"
+    playMajorAndMinorVersion == "2.5"
```
I guess that this change will break Play 2.4 builds because now there is no way for sbt to distinguish whether the build is targeted to Play 2.4 or 2.5.
Would you please confirm that it works for both 2.4 and 2.5?

mumoshu replied:
I guess that this change will break Play 2.4 builds because now there is no way for sbt to distinguish whether the build is targeted to Play 2.4 or 2.5.
Would you please confirm that it works for both 2.4 and 2.5?

mkurz replied:
I removed this method (`isPlay24`) - I can't see where it is used.

mkurz replied:
I removed this method (`isPlay24`) - I can't see where it is used.

cw-kuoka commented on project/Build.scala:
```diff
@@ -49,15 +49,14 @@ object ApplicationBuild extends Build {
       resolvers += "Spy Repository" at "http://files.couchbase.com/maven2",
       resolvers += "Scalaz Bintray Repo"  at "http://dl.bintray.com/scalaz/releases",
-      libraryDependencies += "net.spy" % "spymemcached" % "2.9.0",
+      libraryDependencies += "net.spy" % "spymemcached" % "2.12.0",
```
Thanks for your kindness here!
But we are staying at 2.9.0 because there was once a stability issue in a newer version (as I remember).
I won't agree with bumping it to 2.12.0 without stress testing.

mumoshu replied:
Thanks for your kindness here!
But we are staying at 2.9.0 because there was once a stability issue in a newer version (as I remember).
I won't agree with bumping it to 2.12.0 without stress testing.

cw-kuoka commented on project/Build.scala:
```diff
@@ -49,15 +49,14 @@ object ApplicationBuild extends Build {
       libraryDependencies += "com.typesafe.play" %% "play-cache" % play.core.PlayVersion.current % "provided",
       libraryDependencies += "com.typesafe.play" %% "play-test" % play.core.PlayVersion.current % "provided,test",
-      libraryDependencies += "org.specs2" %% "specs2" % "2.4.15" % "test",
+      libraryDependencies += "org.specs2" %% "specs2" % "3.6.6" % "test",
```
This will break builds for older Play versions since specs2 won't be available for Scala 2.10.x.
I guess that we may be better sticking with 2.4.15 for every build or decide the specs2 version according to the Scala version.

mumoshu replied:
This will break builds for older Play versions since specs2 won't be available for Scala 2.10.x.
I guess that we may be better sticking with 2.4.15 for every build or decide the specs2 version according to the Scala version.

cw-kuoka commented on publish.sh:
```diff
@@ -5,3 +5,5 @@ for play_version in 2.{3.9,4.0}; do
   done
 done
+
+PLAY_VERSION=2.5.0 sbt ++2.11.7 publish-signed
```
LGTM :+1:

mumoshu replied:
LGTM :+1:

cw-kuoka commented on project/plugins.sbt:
```diff
@@ -3,7 +3,7 @@ ivyLoggingLevel := UpdateLogging.Full
 resolvers += "Typesafe repository" at "http://repo.typesafe.com/typesafe/releases/"
 
-val playVersion = scala.util.Properties.envOrElse("PLAY_VERSION", "2.4.0")
+val playVersion = scala.util.Properties.envOrElse("PLAY_VERSION", "2.5.0")
```
LGTM :+1:

mumoshu replied:
LGTM :+1:

cw-kuoka commented on samples/java/play_2.5/activator:
```diff
@@ -288,7 +288,7 @@ declare -a java_args
 declare -a app_commands
```
We really should support older Play versions with the same codebase.
Also, just replacing 2.4 to 2.5 here would stop build against Play 2.4 and someday it'll get broken.
It might be better we create the new sample application under `samples/java/play_2.5` instead of renaming it.

mumoshu replied:
We really should support older Play versions with the same codebase.
Also, just replacing 2.4 to 2.5 here would stop build against Play 2.4 and someday it'll get broken.
It might be better we create the new sample application under samples/java/play_2.5 instead of renaming it.

mkurz commented:
@cw-kuoka If you want you can take my PR and further improve it. I don't know how much time I will find the next few days. Thanks!

mkurz commented:
@mumoshu See my comment before - feel free to take my PR and start working on it. I don't know how much time I find the next few days.

mumoshu commented:
@mkurz Thank you for contribution! Hope I have some time working on it this weekend.

mkurz commented:
@mumoshu I updated this PR.
However there is still a problem because we now we have to use `play.sbt.*`  keys instead of the ones from the package `play.*` in `Build.scala` - e.g. `play.sbt.PlayScala` instead of `play.PlayScala`. These keys have been deprecated and removed now. This makes supporting Play 2.3.9 a bit cumbersome.
See https://github.com/playframework/playframework/blob/2.4.x/framework/src/sbt-plugin/src/main/scala/play/package.scala

mkurz commented:
PIng @mumoshu @cw-kuoka

mumoshu commented:
@mkurz Thanks again for your patience! I have added 2 commits on top of your work.

https://github.com/mumoshu/play2-memcached/compare/c13d3e5032bcaf5d51833455d59e7a1c475fa444...play25upgrade

The issue relating deprecated sbt keys seems to be resolved.
However, Play 2.4 build seems to be failing because it is mistakenly using plugin code for Play 2.3.x located at https://github.com/mumoshu/play2-memcached/tree/master/plugin/src/main/default/com/github/mumoshu/play2/memcached

mkurz commented:
@mumoshu So how should we solve this?

mumoshu commented:
@mkurz Haven't tried yet but maybe modifying both
* the directory names(currently `plugin/src/main/{default,play2.5}`) and
* [the code switching plugin's code-base between targeted Play versions](https://github.com/mumoshu/play2-memcached/blob/play25upgrade/project/Build.scala#L27).
does the job I guess.

mkurz commented:
@mumoshu @cw-kuoka Do you think you can finish this? I am very busy currently...

dhirenpratap commented:
Any update on this?

mkurz commented:
Ping @mumoshu @cw-kuoka

JanRudert commented:
Hi, are there any motivations on finishing this? Maybe drop the Play 2.3 support? It would be better than having no memcached plugin at all in 2.5. Thank you!

mkurz commented:
@mumoshu @cw-kuoka I updated this pull request:

- I added your commits from your `play25upgrade` branch (fixes the sbt deprecations and Java 7 error)
- Activator was removed because its EOL'd already.
- Instead of hacking the plugins code to switch the source I just used symlinks for now, because the plugins code for play 2.4 and play 2.5 is the same anyway.
- I again added the sample for Play 2.4 Scala like on the `master` branch
- I fixed the publish script so it  changes to the `play_2.3` folder when needed (for play 2.3)
- I fixed the hosts hack because it was updated: https://github.com/travis-ci/travis-ci/issues/5227#issuecomment-165131913

Can you please review again? 
After this gets merged I want to work on support for Play 2.6.

Thank you!

mumoshu commented:
Hi @mkurz - Wow, thank you very much for taking your time on this 👍 
As you may have guessed, I've been busy these days and therefore had no chance to work on this.
Your change looks good to me now.
I'll try to publish the new version of play2-memcached for 2.5 asap.
Also, thanks for your willingness to support 2.6. I'd greatly appreciate it 🍺

mkurz commented:
Thanks for merging :+1: You can now also delete the `play25upgrade` branche from this repo.

wwbakker commented:
> After this gets merged I want to work on support for Play 2.6.
> I'll try to publish the new version of play2-memcached for 2.5 asap.
@mkurz @mumoshu That would be awesome, thanks in advance!

jtjeferreira commented:
thx @mkurz @mumoshu for this library. I would like to try it, but I could not find play2-memcached-play25 published anywhere...

mkurz commented:
I hope I find time to upgrade to library to Play 2.6 within the next 2 weeks, I am very busy right now.
However @mumoshu could release the 2.5 support already I think...

mumoshu commented:
Hi, sorry for taking long time but play2-memcached-play25_2.11-0.8.0 is just released to Sonatype.
Perhaps it will be synced to the Central soon.
